chore: update README.md (#22)
synapseweb3/ibc-solidity-contract: README.md

# IBC Solidity For Forcerelay

The IBC solidity contracts for [Axon](https://github.com/axonweb3/axon) is based on [yui-ibc-solidity](https://github.com/hyperledger-labs/yui-ibc-solidity). Most of the instructions, such as registering IBC modules and light client, remain the same.

NOTE: This is pre-beta non-production-quality software, not ready for use in a production environment.

## Limitation

> `yui-ibc-solidity` is designed to be compatible with the IBC core protocol, providing minimal compatibility. In contrast, `ibc-solidity-contract` goes further by providing additional APIs to support Forcerelay's query and storage demands. Consequently, the contract size exceeds the limits imposed by standard EVM-compatible chains, such as Arbitrum, Avalanche, BSC. As a result, this project can only be deployed on Axon, taking advantage of Axon's ability to transcend the EVM's constraints.

Reviewer: ```suggestion
`yui-ibc-solidity` is designed to be compatible with the IBC core protocol, most of its [implementations](https://github.com/hyperledger-labs/yui-ibc-solidity/blob/35814ef028449b10d04f5978f25d88a25064f7aa/docs/architecture.md) are compliant with [ICS](https://github.com/cosmos/ibc) and [ibc-go](https://github.com/cosmos/ibc-go), while `ibc-solidity-contract` goes further by providing additional APIs to support Forcerelay's query and storage demands.

Consequently, [the contract size exceeds the limits](https://github.com/synapseweb3/ibc-solidity-contract/issues/9) imposed by standard EVM-compatible chains, such as Arbitrum, Avalanche and BSC. As a result, this project can only be deployed on Axon, taking advantage of Axon's ability to transcend the EVM's constraints.
```

## Installation and Testing

> First, ensure that, Node.js and Yarn have been already installed:

Reviewer: ```suggestion
First, ensure that, Node.js and Yarn are already installed:
```

```bash
$ git clone https://github.com/synapseweb3/ibc-solidity-contract
$ cd ibc-solidity-contract
$ yarn install
$ yarn compile

> $ yarn test

Reviewer 2: truffle test, but the specified network is not passed as axon-alphanet, so the integration test run is the local network of eth, but I think ci should also run axon-alphanet, because the business logic of the two is different, local eth is used IBCMockHandler, and axon alpha is OwnableIBCHandler, and the ci process of axon-alphanet needs to be supplemented.

> $ yarn test

Reviewer: error Command "test" not found.

Author: I removed it, the answer is [here](https://github.com/synapseweb3/ibc-solidity-contract/pull/22#discussion_r1325877089)

```

> ## Deploy on Axon

Reviewer 2: Do you need to add the operation of axon to deploy self-built nodes and the corresponding ci process?

Reviewer: > Do you need to add the operation of axon to deploy self-built nodes and the corresponding ci process?

It is ok at the current stage.
AXON_HTTP_RPC_URL should be configured in the `.env` file, mentioned in the docs:
https://github.com/synapseweb3/ibc-solidity-contract/blob/b6a93f83f282eef31f9517dcd186099a0fa489ba/README.md?plain=1#L25-L30

Creating a `.env` file in the project root path:

```
AXON_HTTP_RPC_URL=http://<AXON_URL>:<HTTP_PORT>
AXON_WS_RPC_URL=ws://<AXON_URL>:<WS_PORT>
```

Then, use the migration command to deploy the contract and prepare basic settings for Axon:

```bash
$ yarn migrate
```

After deployment, each standalone component in the project has been migrated into Axon and the entry is `OwnableIBCHandler` contract, which accepts operations, including registering IBC module and binding channel port.

## Register Light Client

Registering light client is essential for verifying transactions from the counterparty chain. A mock light client has been registered into OwnableIBCHandler contract during the migration, which acts as an entry for counterparty.

> While the official light clients for verifying Cosmos and CKB transactions are in development, you can use a custom light client by following the [method](https://github.com/synapseweb3/ibc-solidity-contract/blob/master/contracts/core/02-client/ILightClient.sol) below 

Reviewer: ```suggestion
While the official light clients for verifying Cosmos and CKB transactions are in development, you can use a custom light client by the following [method](https://github.com/synapseweb3/ibc-solidity-contract/blob/master/contracts/core/02-client/ILightClient.sol) 
```

```solidity
function registerClient(string calldata clientType, ILightClient client)
```

Note: clientType here can be one of these values: “07-axon”, “07-tendermint”, or “07-ckb4ibc”.

## Bind IBC Module

Before binding the IBC Module, please follow the [MockModule](https://github.com/synapseweb3/ibc-solidity-contract/blob/master/contracts/apps/20-transfer/MockModule.sol) example to create your custom business module contract. Once you've deployed this contract in Axon, you can register its address within the `OwnableIBCHandler` contract by binding it to a channel port:

```solidity
function bindPort(string calldata portId, address moduleAddress)
```

To comply with IBC protocol regulations, a channel cannot function until it is paired with a business module through a port.

## Send ICS20 Packet

In the previous migration step, a standard ICS20 transfer module has already been bound at port `port-0` by default. Run script `scripts/send_packet.js` to send a standard custom ICS20 packet:

```bash

> $ yarn send

Reviewer: ❯ yarn send
yarn run v1.22.19
$ HTTPS_PROXY= HTTP_PROXY= truffle exec './scripts/send_packet.js' --network axon
Using network 'axon'.
AXON_HTTP_RPC_URL=https://rpc-alphanet-axon.ckbapp.dev

send packet from 07-axon-0 to 07-axon-1
TypeError: Cannot set properties of undefined (setting 'onopen')
Truffle v5.11.1 (core: 5.11.1)
Node v18.17.1
error Command failed with exit code 1.

Author: `send` is unworkable because I just use it for debug use, but it's weird if it doesn't show in `README.md`, so I wrote it and consider to complete the `send` command later.

Reviewer: LGTM
